### PR TITLE
Add DecoupledWeightDecay

### PR DESCRIPTION
**Pull request by chengduoZH**
related PR: https://github.com/PaddlePaddle/Paddle/pull/14828

QQ20190328-185200@2x

### CODE REVIEW

File: python/paddle/fluid/optimizer.py [hunk 1]
```diff
@@ -1649,3 +1676,137 @@ def restore(self, executor):
...
+def extend_with_decoupled_weight_decay(base_optimizer):
+    """
+    extend_with_decoupled_weight_decay is a decorator function, it returns an
+    optimizer class with decoupled weight decay. The optimizer of returning will
+    apply weight decay on the optimized parameters with the parameters before
+    optimization, i.e: new_parameter = optimized_parameter - parameter * coeff
+    For more information, please refer: https://arxiv.org/pdf/1711.05101.pdf.
+
+    Args:
+        base_optimizer (Optimizer): The base_optimizer should be a derived class of Optimize.
```
Comment (kuke) [inline python/paddle/fluid/optimizer.py]: `Optimize` -> `Optimizer`

File: python/paddle/fluid/optimizer.py [hunk 2]
```diff
@@ -1649,3 +1676,137 @@ def restore(self, executor):
...
+        return optimize_ops, params_grads
+
+    def __str__(self):
+        return " ".join(["Weight Decay, params:", ",".join(self._params_name)])
+
+
+def extend_with_decoupled_weight_decay(base_optimizer):
+    """
+    extend_with_decoupled_weight_decay is a decorator function, it returns an
+    optimizer class with decoupled weight decay. The optimizer of returning will
```
Comment (kuke) [inline python/paddle/fluid/optimizer.py]: `optimizer of returning` -> `returned optimizer`

**Review by kuke — COMMENTED** [review]
Almost LGTM

File: python/paddle/fluid/optimizer.py (hunk 2 above)
Comment (chengduoZH) [reply python/paddle/fluid/optimizer.py]: Done

File: python/paddle/fluid/optimizer.py (hunk 1 above)
Comment (chengduoZH) [reply python/paddle/fluid/optimizer.py]: Done

**Review by kuke — DISMISSED** [review]
LGTM

**Review by kuke — DISMISSED** [review]
LGTM

**Comment by shanyi15** [comment]
image

**Review by shanyi15 — DISMISSED** [review]
LG comments

**Review by kuke — DISMISSED** [review]

File: python/paddle/fluid/optimizer.py [hunk 3]
```diff
@@ -41,7 +41,8 @@
     'SGDOptimizer', 'MomentumOptimizer', 'AdagradOptimizer', 'AdamOptimizer',
     'AdamaxOptimizer', 'DecayedAdagradOptimizer', 'RMSPropOptimizer',
     'FtrlOptimizer', 'Adadelta', 'ModelAverage', 'LarsMomentum',
-    'LarsMomentumOptimizer', 'DGCMomentumOptimizer'
+    'LarsMomentumOptimizer', 'extend_with_decoupled_weight_decay',
```
Comment (XiaoguangHu01) [inline python/paddle/fluid/optimizer.py]: 这里其他的都是Optimizer，函数功能和命名风格差距较大

Comment (chengduoZH) [reply python/paddle/fluid/optimizer.py]: `extend_with_decoupled_weight_decay ` 返回的也是一个Optimizer的。

**Review by shanyi15 — APPROVED** [review]
LG comments

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**